rclcpp_components/rclcpp_components_template_node.cmake.in

```diff
@@ -0,0 +1,74 @@
+# Copyright 2019 Open Source Robotics Foundation, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# usage: rclcpp_components_register_node(
+#              <target> PLUGIN <component> EXECUTABLE <node>)
+#
+# Register an rclcpp component with the ament
+# resource index and create an executable.
+#
+# :param target: the shared library target
+# :type target: string
+# :param ARGS: the plugin name and node name for executable
+# :type ARGS: list of strings
+# :param component: the component plugin name
+# :type component: string
+# :param node: the node's executable name
+# :type node: string
+#
+macro(rclcpp_components_register_node target)
+  cmake_parse_arguments(
+    ARGS
+    ""
+    "PLUGIN;EXECUTABLE"
+    ""
+    ${ARGN})
+  set(component ${ARGS_PLUGIN})
+  set(node ${ARGS_EXECUTABLE})
+  _rclcpp_components_register_package_hook()
+  set(_path "lib")
+  if(WIN32)
+    set(_path "bin")
+    set(LIBRARY_NAME ${target}.dll)
+  elseif(APPLE)
+    set(LIBRARY_NAME lib${target}.dylib)
+  else()
+    set(LIBRARY_NAME lib${target}.so)
+  endif()
+  set(_RCLCPP_COMPONENTS__NODES
+    "${_RCLCPP_COMPONENTS__NODES}${component};${_path}/$<TARGET_FILE_NAME:${target}>\n")
+  set(CLASS_NAME ${component})
+  set(build_dir @install_dir@)
+  configure_file(${build_dir}/node_main.cpp.in
+    ${build_dir}/node_main_${node}.cpp @ONLY)
```

Comment:
I don't think the resultant code should be placed in `rclcpp_components` install directory, but rather the calling code's build directory. Perhaps this should contain a sub-directory named `rclcpp_components` to avoid potential name clashes, e.g. `${PROJECT_BINARY_DIR}/rclcpp_components/node_main_${node}.cpp`